feat(calendar): add calendar widget

sophacles opened the pull request:
A calendar widget that with some display options and styling for individual dates.

This widget displays the month provided to it in the constructor. (see docs and the example for more info).

One thing to note: This pulls in a new dependency, the [time](https://crates.io/crates/time) crate. Since this adds a dependency, it might be worth putting the widget behind a feature flag - I'm happy to do so if that's desired.

joshka reviewed (approved):
Looks great! Some minor comments only

sophacles commented on src/widgets/calendar.rs:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    /// Display a header containing weekday abbreviations
+    pub fn show_weekdays(mut self, style: Style) -> Self {
```
Good idea. Updated.

sophacles commented on examples/calendar.rs:
```diff
@@ -0,0 +1,256 @@
+        Date::from_calendar_date(current_year, Month::April, 22).unwrap(),
+        holiday_style,
+    );
+    // star wars day
+    list.add(
+        Date::from_calendar_date(current_year, Month::May, 4).unwrap(),
+        holiday_style,
+    );
```
Thanks!

sophacles commented on src/widgets/calendar.rs:
```diff
@@ -0,0 +1,214 @@
+
+use time::{Date, Duration, OffsetDateTime};
+
+/// Display a month calendar for the month containing `display_date`
+pub struct Calendar<'a, S: DateStyler> {
```
It's a good idea. What I'd really like is to set it up for use like:

```
use ratatui::widgets::calendars::{Month, Week /* and so on ...*/};
```
But that deviates from the current pattern a lot.

Thoughts?

joshka replied:
> `use ratatui::widgets::calendars::{Month, Week /* and so on ...*/};`

`widgets::calendars::{Month,Week}` makes 100% sense to me. Great idea.

Btw, the lint error just looks like it's for the commit message (which will go away with a squash).

sophacles replied:
Linter errors were taken care of by squashing, thanks for the suggestions!

joshka reviewed (approved):
Changes for header LGTM.

I'm not sure whether the `time` dependency will be an actual problem for anyone (I'd guess it's probably a fundamental enough crate that probably not). The bigger issue on that idea though is that if you do plan to move to using chrono down the line, perhaps feature flagging on whether to use time or chrono calendars could be worth doing up front. I don't have any strong leanings either way on this. It could be best to just leave this until it's an issue.

I'd lean towards adding the change to `calendars::Month` in this PR the name just seems more correct and changing the widget names down the line would break more dependents.

Slapping an approve as no big deal if you don't make that change either.

sayanarijit commented:
Should we put calendar and other similar new-dependency-introducing widgets behind feature flags? @orhun @mindoodoo

sayanarijit commented:
> Hmm, what would you name the feature ?

`widget-calendar`?

sophacles commented:
I put in the machinery for the feature and named it `widget-calendar`. It's easy to change if something else is decided.

orhun commented:
I didn't understand why you went with `dep:time` instead of `time`, why that is required?

sophacles commented:
> why is that required?

Apparently it isn't :). I'm still learning about some of these Cargo directives and I think I misunderstood some of what I read.

Pushed a fixup

orhun commented:
Just tested and it works well. GJ! 🎉

One question though, is it possible to add tests for this?

sophacles commented:
I can add tests (i think). I'll see what I can get in place over the weekend.

sophacles commented:
I've added some tests for the widget, as well as for the date styler.

One thing to note, I'm not sure if CI will actually run the tests, because I don't know if it builds with default features or does something "smart". I added a feature `all-widgets` in my latest commit, which currently just specs `widget-calendar`, but as we add other widgets behind feature flags, I presume it will get more entries. It also allows for a simpler testing with `cargo test --features all-widgets`

orhun commented:
I think you can update `Makefile.toml` like this for running tests for `all-widgets` feature:

```diff
diff --git a/Makefile.toml b/Makefile.toml
index 8afba0a..db9354f 100644
--- a/Makefile.toml
+++ b/Makefile.toml
@@ -100,11 +100,11 @@ args = [
 ]
 
 [tasks.test-crossterm]
-env = { TUI_FEATURES = "serde,crossterm" }
+env = { TUI_FEATURES = "serde,crossterm,all-widgets" }
 run_task = "test"
 
 [tasks.test-termion]
-env = { TUI_FEATURES = "serde,termion" }
+env = { TUI_FEATURES = "serde,termion,all-widgets" }
 run_task = "test"
 
 [tasks.test]
```

sophacles commented:
Thanks @orhun - that looks like it did the trick, I can calendar tests being run now.

mindoodoo reviewed (approved):
This looks awesome man !
